examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2893 @@
+  // \boldsymbol{\nu}_i := \sum_{T \subset \text{supp}(\phi_i)}
+  // \sum_{F \subset \partial T \cap \partial \Omega}
+  // \sum_{\mathbf{x}_{q,F}} \nu(\mathbf{x}_{q,F})
+  // \phi_i(\mathbf{x}_{q,F})
+  // @f}
+  //
+  // here $T$ denotes elements,
+  // $\text{supp}(\phi_i)$ the support of the shape function $\phi_i$,
+  // $F$ are faces of the element $T$, and $\mathbf{x}_{q,F}$
+  // are quadrature points on such face. Note that this formula for
+  // $\widehat{\boldsymbol{\nu}}_i$ is nothing else than some form of
+  // weighted averaging. Other more sophisticated definitions for $\nu_i$
+  // are possible but none of them have much influence in theory or practice.
+
+  template <int dim>
+  void OfflineData<dim>::assemble()
+  {
+    lumped_mass_matrix = 0.;
+    norm_matrix        = 0.;
+    for (auto &matrix : cij_matrix)
+      matrix = 0.;
+    for (auto &matrix : nij_matrix)
+      matrix = 0.;
+
+    unsigned int dofs_per_cell = discretization->finite_element.dofs_per_cell;
+    unsigned int n_q_points    = discretization->quadrature.size();
+
+    // What follows is the implementation of the scratch data required by
+    // WorkStream
+
+    MeshWorker::ScratchData<dim> scratch_data(
+      discretization->mapping,
+      discretization->finite_element,
+      discretization->quadrature,
+      update_values | update_gradients | update_quadrature_points |
+        update_JxW_values,
+      discretization->face_quadrature,
+      update_normal_vectors | update_values | update_JxW_values);
+
+    {
+      TimerOutput::Scope t(
+        computing_timer,
+        "offline_data - assemble lumped mass matrix, and c_ij");
+
+      const auto local_assemble_system = [&](const auto &cell,
+                                             auto &      scratch,
+                                             auto &      copy) {
+        auto &is_artificial             = copy.is_artificial;
+        auto &local_dof_indices         = copy.local_dof_indices;
+        auto &local_boundary_normal_map = copy.local_boundary_normal_map;
+        auto &cell_lumped_mass_matrix   = copy.cell_lumped_mass_matrix;
+        auto &cell_cij_matrix           = copy.cell_cij_matrix;
+
+        is_artificial = cell->is_artificial();
+        if (is_artificial)
+          return;
+
+        local_boundary_normal_map.clear();
+        cell_lumped_mass_matrix.reinit(dofs_per_cell, dofs_per_cell);
+        for (auto &matrix : cell_cij_matrix)
+          matrix.reinit(dofs_per_cell, dofs_per_cell);
+
+        const auto &fe_values = scratch.reinit(cell);
+
+        local_dof_indices.resize(dofs_per_cell);
+        cell->get_dof_indices(local_dof_indices);
+
+        std::transform(local_dof_indices.begin(),
+                       local_dof_indices.end(),
+                       local_dof_indices.begin(),
+                       [&](auto index) {
+                         return partitioner->global_to_local(index);
```

Comment:
This is the same code piece I mentioned above might be nice to have inside `partitioner`. Since it appears a second time here, it might indeed be nice to have that in the `Partitioner` class.